### PR TITLE
server/notifications: rewrite notification foundations

### PR DESCRIPTION
**Pull request by zegl**
- Simplify notification types, move all payloads, templates, etc to the same place
- Remove deduplication key. If you want to send a notification, it should already be deduplicated
- Support sending emails to anonymous pledgers

- Email copy (rendered as HTML): https://github.com/polarsource/polar/tree/emails-v2/server/tests/notifications/testdata
- In app copy (storybook): https://www.chromatic.com/component?appId=646dd122a99fb9a2fe564ec3&csfId=organisms-notification&buildNumber=6&k=646f2789de280316615e5e1e-1200-interactive-true&h=9&b=-1

## Personal settings (self user org not installed)

Screenshot 2023-05-25 at 11 06 01

## Org settings (non-user org)

Screenshot 2023-05-25 at 11 06 06

## Self user org

Screenshot 2023-05-25 at 11 09 42

## TODO (in future PRs)

- [ ] connect email sending to events
- [ ] remove old email preferences

### CODE REVIEW

File: server/polar/notifications/notification.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+from dataclasses import dataclass
+from jinja2.nativetypes import NativeEnvironment
+from jinja2 import StrictUndefined
+
+from polar.models.user import User
+from pydantic import BaseModel, Extra
+
+
+# @dataclass
+class NotificationType(BaseModel):
```
Comment (sultaniman) [inline server/polar/notifications/notification.py]: Maybe something like `NotificationBase` is more clear and communicates purpose as well while `NotificationType` implies it is an actual type?

File: server/polar/receivers/pledges.py [hunk 2]
```diff
@@ -32,15 +39,42 @@ async def mark_pledges_pending_on_issue_close(
 async def pledge_created_notification(pledge: Pledge, session: AsyncSession):
     issue = await issue_service.get_by_id(session, pledge.issue_id)
     if not issue:
+        log.error("pledge_created_notification.no_issue_found")
```
Comment (sultaniman) [inline server/polar/receivers/pledges.py:42]: Maybe logging `issue_id` `org_id` will be useful for debugging?

File: server/polar/receivers/pledges.py [hunk 3]
```diff
@@ -32,15 +39,42 @@ async def mark_pledges_pending_on_issue_close(
...
+    if not org:
+        log.error("pledge_created_notification.no_org_found")
         return
 
+    repo = await repository_service.get(session, issue.repository_id)
+    if not repo:
+        log.error("pledge_created_notification.no_repo_found")
+        return
+
+    issue_url = f"https://github.com/{org.name}/{repo.name}/issues/{issue.number}"
```
Comment (sultaniman) [inline server/polar/receivers/pledges.py:55]: shall this be in settings instead?

File: server/polar/notifications/notification.py (hunk 1 above)
Comment (zegl) [reply server/polar/notifications/notification.py]: This is a good suggestion, thanks! :-)

**Comment by zegl** [comment]
@sultaniman Not at all! It was good to get some extra eyes on what’s happening, and it’s cool to see that you’re getting yourself this deep into Polar! ❤️

**Pull request merged**